In [83]:
import pandas as pd

In [84]:
df = pd.read_csv('VOX_Data_Fixed_With_Seg_Counts_New_For_Testing.csv')

In [85]:
df = df.iloc[0:3149]
df.shape

(3149, 27)

In [86]:
df = df[df['category_key'] == 'Restaurants']
df = df[df['offer_title']=='Spend $25 and get $5 Cashback']

In [87]:
df.shape

(1990, 27)

In [88]:
df.head()

date     transaction_date                         cardholder_id  \
0  5/30/2024 15:30  2024-05-27T08:06:26  fb09e602-bfa8-4e82-8342-6b5a29cc75dc   
1  5/30/2024 15:30  2024-05-27T01:47:16  79bcfb7b-f09d-4aa4-8f8c-7c9b743235f4   
3  5/30/2024 15:15  2024-05-27T08:24:39  d47b7a1f-6a6b-44f9-b343-7ef3fd175e60   
4  5/30/2024 15:15  2024-05-27T10:03:43  aae0cc03-49e7-4ce2-adf5-933e42c7075f   
5  5/30/2024 15:15  2024-05-27T07:18:02  a0e829c1-f05b-49d4-bf59-24056b0959ee   

                                card_id card_segment offer_segment  \
0  ce00cec1-fa01-ef11-a574-b8cef6052491        LOYAL           NaN   
1  73c592cf-2802-ef11-a574-b8cef6052491        LOYAL           NaN   
3  858d17f4-0a02-ef11-a574-b8cef6052491        LOYAL           NaN   
4  349c294b-0e0d-ef11-abbe-043f72dce83d        LOYAL           NaN   
5  b0408482-3002-ef11-a574-b8cef6052491        LOYAL           NaN   

   transaction_id transaction_currency  transaction_currency_code  \
0    4.641483e+14                  AUD                       36.0   
1    3.841481e+14                  AUD                       36.0   
3    3.841483e+14                  AUD                       36.0   
4    5.841484e+14                  AUD                       36.0   
5    3.841483e+14                  AUD                       36.0   

   transaction_amount  ...                    offer_title  ruleset_id  \
0               27.04  ...  Spend $25 and get $5 Cashback       291.0   
1               30.68  ...  Spend $25 and get $5 Cashback       291.0   
3               26.89  ...  Spend $25 and get $5 Cashback       291.0   
4               51.97  ...  Spend $25 and get $5 Cashback       291.0   
5              133.20  ...  Spend $25 and get $5 Cashback       291.0   

  redemption_limit discount_type fixed_value minimum_spend  \
0              NaN         FIXED         5.0          25.0   
1              NaN         FIXED         5.0          25.0   
3              NaN         FIXED         5.0          25.0   
4              NaN         FIXED         5.0          25.0   
5              NaN         FIXED         5.0          25.0   

   reward_every_transaction reward_frequency  minimum_transactions  \
0                        NO             ONLY                   1.0   
1                        NO             ONLY                   1.0   
3                        NO             ONLY                   1.0   
4                        NO             ONLY                   1.0   
5                        NO             ONLY                   1.0   

   casback_amount  
0             5.0  
1             5.0  
3             5.0  
4             5.0  
5             5.0  

[5 rows x 27 columns]

In [89]:
df = df.drop_duplicates('card_id')

In [90]:
df.shape

(1864, 27)

In [91]:
df.head()

date     transaction_date                         cardholder_id  \
0  5/30/2024 15:30  2024-05-27T08:06:26  fb09e602-bfa8-4e82-8342-6b5a29cc75dc   
1  5/30/2024 15:30  2024-05-27T01:47:16  79bcfb7b-f09d-4aa4-8f8c-7c9b743235f4   
3  5/30/2024 15:15  2024-05-27T08:24:39  d47b7a1f-6a6b-44f9-b343-7ef3fd175e60   
4  5/30/2024 15:15  2024-05-27T10:03:43  aae0cc03-49e7-4ce2-adf5-933e42c7075f   
5  5/30/2024 15:15  2024-05-27T07:18:02  a0e829c1-f05b-49d4-bf59-24056b0959ee   

                                card_id card_segment offer_segment  \
0  ce00cec1-fa01-ef11-a574-b8cef6052491        LOYAL           NaN   
1  73c592cf-2802-ef11-a574-b8cef6052491        LOYAL           NaN   
3  858d17f4-0a02-ef11-a574-b8cef6052491        LOYAL           NaN   
4  349c294b-0e0d-ef11-abbe-043f72dce83d        LOYAL           NaN   
5  b0408482-3002-ef11-a574-b8cef6052491        LOYAL           NaN   

   transaction_id transaction_currency  transaction_currency_code  \
0    4.641483e+14                  AUD                       36.0   
1    3.841481e+14                  AUD                       36.0   
3    3.841483e+14                  AUD                       36.0   
4    5.841484e+14                  AUD                       36.0   
5    3.841483e+14                  AUD                       36.0   

   transaction_amount  ...                    offer_title  ruleset_id  \
0               27.04  ...  Spend $25 and get $5 Cashback       291.0   
1               30.68  ...  Spend $25 and get $5 Cashback       291.0   
3               26.89  ...  Spend $25 and get $5 Cashback       291.0   
4               51.97  ...  Spend $25 and get $5 Cashback       291.0   
5              133.20  ...  Spend $25 and get $5 Cashback       291.0   

  redemption_limit discount_type fixed_value minimum_spend  \
0              NaN         FIXED         5.0          25.0   
1              NaN         FIXED         5.0          25.0   
3              NaN         FIXED         5.0          25.0   
4              NaN         FIXED         5.0          25.0   
5              NaN         FIXED         5.0          25.0   

   reward_every_transaction reward_frequency  minimum_transactions  \
0                        NO             ONLY                   1.0   
1                        NO             ONLY                   1.0   
3                        NO             ONLY                   1.0   
4                        NO             ONLY                   1.0   
5                        NO             ONLY                   1.0   

   casback_amount  
0             5.0  
1             5.0  
3             5.0  
4             5.0  
5             5.0  

[5 rows x 27 columns]

In [92]:
print('Transaction amount sum is ',df['transaction_amount'].sum())
print('minimum spend amount sum is ',df['minimum_spend'].sum())
print('cashback amount sum is ',df['casback_amount'].sum())

Transaction amount sum is  94505.16
minimum spend amount sum is  46600.0
cashback amount sum is  9320.0


**That means Menulog spent 9765 dollars for 651 customers to earn minimum amount of 19530 dollars and maximum amount of 36898 dollars**

In [93]:
df_RFM = pd.read_csv('RFM Values For Card_ID.csv')

In [94]:
df_RFM = df_RFM[['card_id','Recency','Frequency','Monetary','Cluster','wallet_share_by_category']]

In [95]:
df_RFM.head()

card_id  Recency  Frequency  Monetary  \
0  27348eaa-1902-ef11-a574-b8cef6052491       23          1     55.15   
1  1431d2e3-0c02-ef11-a574-b8cef6052491       22          1    122.95   
2  709b010b-1b02-ef11-a574-b8cef6052491       23          4    192.87   
3  13079fad-3bf3-ee11-a574-b8cef6052491       23          1     83.44   
4  e077c288-2902-ef11-a574-b8cef6052491       23          1    143.43   

   Cluster  wallet_share_by_category  
0        2                     55.15  
1        2                    122.95  
2        0                    192.87  
3        2                     83.44  
4        2                    143.43

In [96]:
df_RFM = df_RFM.drop_duplicates('card_id')
df_RFM.shape

(3864, 6)

In [97]:

# Assuming df and df_RFM are already defined
# Merge the dataframes on 'card_id'
df_combined = pd.merge(df, df_RFM, on='card_id', how='left')


In [98]:
df_combined.head()

date     transaction_date                         cardholder_id  \
0  5/30/2024 15:30  2024-05-27T08:06:26  fb09e602-bfa8-4e82-8342-6b5a29cc75dc   
1  5/30/2024 15:30  2024-05-27T01:47:16  79bcfb7b-f09d-4aa4-8f8c-7c9b743235f4   
2  5/30/2024 15:15  2024-05-27T08:24:39  d47b7a1f-6a6b-44f9-b343-7ef3fd175e60   
3  5/30/2024 15:15  2024-05-27T10:03:43  aae0cc03-49e7-4ce2-adf5-933e42c7075f   
4  5/30/2024 15:15  2024-05-27T07:18:02  a0e829c1-f05b-49d4-bf59-24056b0959ee   

                                card_id card_segment offer_segment  \
0  ce00cec1-fa01-ef11-a574-b8cef6052491        LOYAL           NaN   
1  73c592cf-2802-ef11-a574-b8cef6052491        LOYAL           NaN   
2  858d17f4-0a02-ef11-a574-b8cef6052491        LOYAL           NaN   
3  349c294b-0e0d-ef11-abbe-043f72dce83d        LOYAL           NaN   
4  b0408482-3002-ef11-a574-b8cef6052491        LOYAL           NaN   

   transaction_id transaction_currency  transaction_currency_code  \
0    4.641483e+14                  AUD                       36.0   
1    3.841481e+14                  AUD                       36.0   
2    3.841483e+14                  AUD                       36.0   
3    5.841484e+14                  AUD                       36.0   
4    3.841483e+14                  AUD                       36.0   

   transaction_amount  ... minimum_spend  reward_every_transaction  \
0               27.04  ...          25.0                        NO   
1               30.68  ...          25.0                        NO   
2               26.89  ...          25.0                        NO   
3               51.97  ...          25.0                        NO   
4              133.20  ...          25.0                        NO   

  reward_frequency minimum_transactions casback_amount Recency  Frequency  \
0             ONLY                  1.0            5.0     NaN        NaN   
1             ONLY                  1.0            5.0    27.0        1.0   
2             ONLY                  1.0            5.0     NaN        NaN   
3             ONLY                  1.0            5.0     NaN        NaN   
4             ONLY                  1.0            5.0    36.0        1.0   

  Monetary  Cluster  wallet_share_by_category  
0      NaN      NaN                       NaN  
1    46.89      2.0                     46.89  
2      NaN      NaN                       NaN  
3      NaN      NaN                       NaN  
4   133.80      1.0                    133.80  

[5 rows x 32 columns]

In [99]:
# Remove rows where 'frequency' column is NaN
df_combined = df_combined.dropna(subset=['Frequency'])


In [100]:
df_combined.shape

(597, 32)

In [101]:
df_combined.head()

date     transaction_date                         cardholder_id  \
1  5/30/2024 15:30  2024-05-27T01:47:16  79bcfb7b-f09d-4aa4-8f8c-7c9b743235f4   
4  5/30/2024 15:15  2024-05-27T07:18:02  a0e829c1-f05b-49d4-bf59-24056b0959ee   
5  5/30/2024 15:15  2024-05-27T08:33:00  1db13f51-b162-4c7a-b7a0-d1fa58b4931d   
6  5/30/2024 15:00  2024-05-27T07:51:17  a6da8ff9-cb30-44f2-a876-a0331eae6a39   
8  5/30/2024 14:45  2024-05-27T01:36:49  dec19607-3d37-4d93-9172-1d322c1c648b   

                                card_id card_segment offer_segment  \
1  73c592cf-2802-ef11-a574-b8cef6052491        LOYAL           NaN   
4  b0408482-3002-ef11-a574-b8cef6052491        LOYAL           NaN   
5  87814eec-fb01-ef11-a574-b8cef6052491        LOYAL           NaN   
6  9e062fa0-fd01-ef11-a574-b8cef6052491        LOYAL           NaN   
8  be2aa0df-2d02-ef11-a574-b8cef6052491        LOYAL           NaN   

   transaction_id transaction_currency  transaction_currency_code  \
1    3.841481e+14                  AUD                       36.0   
4    3.841483e+14                  AUD                       36.0   
5    3.841483e+14                  AUD                       36.0   
6    5.841483e+14                  AUD                       36.0   
8    3.041481e+14                  AUD                       36.0   

   transaction_amount  ... minimum_spend  reward_every_transaction  \
1               30.68  ...          25.0                        NO   
4              133.20  ...          25.0                        NO   
5               55.81  ...          25.0                        NO   
6               65.79  ...          25.0                        NO   
8               43.69  ...          25.0                        NO   

  reward_frequency minimum_transactions casback_amount Recency  Frequency  \
1             ONLY                  1.0            5.0    27.0        1.0   
4             ONLY                  1.0            5.0    36.0        1.0   
5             ONLY                  1.0            5.0    29.0        2.0   
6             ONLY                  1.0            5.0    32.0        1.0   
8             ONLY                  1.0            5.0    37.0        1.0   

  Monetary  Cluster  wallet_share_by_category  
1    46.89      2.0                     46.89  
4   133.80      1.0                    133.80  
5   116.82      2.0                    116.82  
6    33.90      1.0                     33.90  
8    29.13      1.0                     29.13  

[5 rows x 32 columns]

In [102]:
print('Transaction amount sum is ',df_combined['transaction_amount'].sum())
print('minimum spend amount sum is ',df_combined['minimum_spend'].sum())
print('cashback amount sum is ',df_combined['casback_amount'].sum())

Transaction amount sum is  29550.399999999998
minimum spend amount sum is  14925.0
cashback amount sum is  2985.0


**That means Menulog spent 2985 dollars for 597 customers to earn minimum amount of 14925.0 dollars and maximum amount of 29550.399999999998 dollars**

In [104]:
df_for_testing = df_combined[['Recency','Frequency','Monetary','Cluster','wallet_share_by_category']]

In [105]:
df_for_testing.head()

Recency  Frequency  Monetary  Cluster  wallet_share_by_category
1     27.0        1.0     46.89      2.0                     46.89
4     36.0        1.0    133.80      1.0                    133.80
5     29.0        2.0    116.82      2.0                    116.82
6     32.0        1.0     33.90      1.0                     33.90
8     37.0        1.0     29.13      1.0                     29.13

In [80]:
df_for_testing['minimum_spend']=25
df_for_testing['seg_unique_user_count']=597
df_for_testing['casback_amount']=5
df_for_testing['category_key_Fashion & Retail']=False
df_for_testing['category_key_Restaurants']=True
df_for_testing['category_key_Retailers']=False

/tmp/ipykernel_2786/2720231731.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_for_testing['minimum_spend']=25
/tmp/ipykernel_2786/2720231731.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_for_testing['seg_unique_user_count']=597
/tmp/ipykernel_2786/2720231731.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stabl

In [81]:
df_for_testing.head()

Recency  Frequency  Monetary  Cluster  wallet_share_by_category  \
1     27.0        1.0     46.89      2.0                     46.89   
4     36.0        1.0    133.80      1.0                    133.80   
5     29.0        2.0    116.82      2.0                    116.82   
6     32.0        1.0     33.90      1.0                     33.90   
8     37.0        1.0     29.13      1.0                     29.13   

   minimum_spend  seg_unique_user_count  casback_amount  \
1             25                    597               5   
4             25                    597               5   
5             25                    597               5   
6             25                    597               5   
8             25                    597               5   

   category_key_Fashion & Retail  category_key_Restaurants  \
1                          False                      True   
4                          False                      True   
5                          False                      True   
6                          False                      True   
8                          False                      True   

   category_key_Retailers  
1                   False  
4                   False  
5                   False  
6                   False  
8                   False

In [82]:
df_for_testing.to_csv('Final Dataset to Test using the Trained Model.csv',index=False)